hashintel/hash
# Task-Executor

## 🌟 What is the purpose of this PR?

In the near-future we are going to need to be able to execute tasks outside of the backend server that currently serves HASH. Doing so will require us to be able to run Python scripts, and possibly other tools as well.

This PR introduces a short-term solution, which is a lightweight server which uses subprocesses to execute scripts. Those scripts are bundled with the server into a docker image which we will be able to deploy as a resource within our k8s cluster.

**Later on this work will be replaced by open-source tooling to provide proper task scheduling, orchestration, resilience, etc.** 

## ⚠️ Known issues

This is meant as a quick-fix to our needs, it will not satisfy many of the longer term considerations to do with retry logic, orchestration, scale, etc.

## 🐾 Next steps

- Replace the Python demo route with a [Singer tap](https://www.singer.io/)
- Setup a codepath in the HASH backend to communicate with the task-executor

## 🔍 What does this change?

- Adds a new `task-executor` package
- Adds a Python script `demo` which returns a string
- Adds a Dockerfile to build the `task-executor` package into an image

### 📜 Does this require a change to the docs?

- This isn't meant to be used outside of some on-going development work at the moment, the accompanying documentation has been updated to explain _how_ to use it, but until it _should_ be used there's no need for more documentation.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202073043981052/f) _(internal)_

## 🛡 What tests cover this?

- None

## ❓ How to test this?

1.  Follow the READMEs on how to setup the package and build the docker image
2. Run the docker image and make sure to forward the 5010 port
3. Go to localhost:${PORT} and query `/python`
4. ???
5. Profit

## 📹 Demo

<img width="311" alt="image" src="https://user-images.githubusercontent.com/25749103/161806814-2ba18093-193b-4bc1-a33c-063718079b43.png">
